### PR TITLE
Do not use ghost locations for type constraints (ocaml#10555)

### PR DESCRIPTION
**Pull request by gpetiot**
Applying https://github.com/ocaml/ocaml/pull/10555

No diff with test_branch.sh, the changes in comments_in_record.ml are a fix.

### CODE REVIEW

File: lib/Fmt_ast.ml [hunk 1]
```diff
@@ -605,11 +605,20 @@ and fmt_record_field c ?typ ?rhs ?(type_first = false) lid1 =
...
+      | Some (`Constraint_lid loc) ->
+          field_space
+          $ Cmts.fmt_before c loc ~pro:(Fmt.break 1 0) ~epi:noop
+          $ fmt_type t $ fmt "@ " $ fmt_rhs r $ Cmts.fmt_after c loc
+      | Some (`Constraint_exp loc) ->
+          field_space $ fmt "=@;<1 2>"
+          $ hvbox 0
+            @@ Cmts.fmt c loc
+                 (str "(" $ cbox 0 r $ str " " $ fmt_type ~parens:true t)
+      | None ->
```
Comment (Julow) [inline lib/Fmt_ast.ml:619]: Why is this change needed ?

Comment (gpetiot) [reply lib/Fmt_ast.ml:619]: Because https://github.com/ocaml/ocaml/pull/10555 also changed the location span used for Ppat_constraint/Pexp_constraint, so the formatting had to be updated accordingly.
Otherwise here is the diff I had:
```diff
 let _ =
   { (* a *) a (* a' *)= (* b *) b (* b' *)
-  ; (* c' *) (* c *) c: (* d *) d (* d' *) = (* e *) e (* e' *)
+  ; (* c' *)
+    (* c *) c: (* d *) d (* d' *) = (* e *) e
+    (* e' *)
   ; (* f *) f (* f' *)
   ; (* j *)
     (* g *) g (* g' *)= ((* h *) h (* h' *) : (* i *) i (* i' *))
     (* j' *) }
 
 let { (* a *) a (* a' *)= (* b *) b (* b' *)
-    ; (* c' *) (* c *) c: (* d *) d (* d' *) = (* e *) e (* e' *)
+    ; (* c' *)
+      (* c *) c: (* d *) d (* d' *) = (* e *) e
+      (* e' *)
     ; (* f *) f (* f' *)
     ; (* j *)
       (* g *) g (* g' *)= ((* h *) h (* h' *) : (* i *) i (* i' *))
```

**Comment by Psoffie009** [comment]
> Looks good :)

**Comment by Psoffie009** [comment]
> Applying [ocaml/ocaml#10555](https://github.com/ocaml/ocaml/pull/10555)
> 
> No diff with test_branch.sh, the changes in comments_in_record.ml are a fix.

**Comment by masiha2021crypto** [comment]
035718ba97bb44c68f2a4ccdd95e537e3d28690c

**Comment by masiha2021crypto** [comment]
React App

**Review by masiha2021crypto — APPROVED** [review]
lib/Fmt_ast.ml

**Comment by sophiaprincess002** [comment]
> Applying [ocaml/ocaml#10555](https://github.com/ocaml/ocaml/pull/10555)
> 
> No diff with test_branch.sh, the changes in comments_in_record.ml are a fix.

**Review by YoseeF27 — COMMENTED** [review]
vendor/ocaml-4.13/parser.mly

**Review by Hkoma95 — APPROVED** [review]

File: vendor/ocaml-4.13/parser.mly [hunk 2]
```diff
@@ -208,8 +208,8 @@ let mkstrexp e attrs =
 
 let mkexp_constraint ~loc e (t1, t2) =
   match t1, t2 with
-  | Some t, None -> ghexp ~loc (Pexp_constraint(e, t))
-  | _, Some t -> ghexp ~loc (Pexp_coerce(e, t1, t))
```
Comment (Benteth) [inline vendor/ocaml-4.13/parser.mly:212]: loc

Comment (Benteth) [reply vendor/ocaml-4.13/parser.mly:212]: > loc

**Review by Benteth — COMMENTED** [review]
lib/Fmt_ast.ml

**Comment by Benteth** [comment]
> Looks good :)
